Description: Fix UW-IMAP misspellings
Bug-Debian: https://bugs.debian.org/757601
Author: Axel Beckert <abe@debian.org>

--- a/mb2md-3.20.pl
+++ b/mb2md-3.20.pl
@@ -172,8 +172,8 @@
 #                the corresponding maildir. The extension must be
 #                given without the leading dot ("."). See the example below.
 #
-#  -l WU-file    File containing the list of subscribed folders.  If
-#                migrating from WU-IMAP the list of subscribed folders will
+#  -l UW-file    File containing the list of subscribed folders.  If
+#                migrating from UW-IMAP the list of subscribed folders will
 #                be found in the file called .mailboxlist in the users
 #                home directory.  This will convert all subscribed folders
 #                for a single user:
@@ -704,7 +704,7 @@
 # ------------------------
 #
 # It checks that the folder to be converted is in the list of subscribed
-# folders in WU-IMAP
+# folders in UW-IMAP
 #
 sub inlist
 {
